### PR TITLE
[Fix #11277] Add new `Style/ConcatArrayLiterals` cop

### LINKED ISSUE #11277: Cop idea: recommend `push` over `concat` for single element lists
I often see new Ruby devs use `concat` to add an element to a list. This cop would recommend `push` (or `<<` depending on configuration).

```ruby
# bad
list.concat [item]
# good
list.push item
list << item
```

It also happens to be about 3x faster (10M iterations here):
```
        user     system      total        real
concat  1.055298   0.037218   1.092516 (  1.092645)
push    0.368790   0.017711   0.386501 (  0.386501)
```

### PR DESCRIPTION
**Pull request by koic**
Fixes #11277.

## Summary

Enforces the use of `Array#push(item)` instead of `Array#concat([item])` to avoid redundant array literal.

## Safety

This cop is unsafe, as it can produce false positives if the receiver is not an `Array` object.

## Examples

```ruby
# bad
list.concat([item])

# good
list.push(item)
```

If the array literal has multiple elements like `item.concat([foo, bar])`, replace it with `item.push(foo, bar)`. `<<` is not allowed multiple arguments. So for autocorrection choose `push`.

## Benchmark

```ruby
# concat_array_literals.rb
require 'benchmark/ips'

ary = [:a, :b, :c]

Benchmark.ips do |x|
  x.report('concat') { ary.concat([:d, :e, :f]) }
  x.report('push')   { ary.push(:d, :e, :f) }

  x.compare!
end
```

```console
% ruby concat_array_literals.rb
Warming up --------------------------------------
              concat   616.057k i/100ms
                push     1.099M i/100ms
Calculating -------------------------------------
              concat      7.382M (±14.5%) i/s -     35.115M in   5.388710s
                push     11.604M (± 2.9%) i/s -     58.241M in   5.023063s

Comparison:
                push: 11604292.6 i/s
              concat:  7382414.4 i/s - 1.57x  (± 0.00) slower
```

## Additional Infromation

Percent array literals are not supported because the complexity of autocorrection logic.

And I think this cop can be placed in Style department, not Performance department. Because there doesn't seem to be much risk of lowering readability.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by dvandersluis** [comment]
Should this be in `Performance` instead of `Style`?

**Comment by koic** [comment]
Quote from this PR description:

> And I think this cop can be placed in Style department, not Performance department. Because there doesn't seem to be much risk of lowering readability.

I think performance is secondary and can be put in `Style`. For example, this style has the advantage of eliminating the use of redundant array literals. That said, of course any feedback on this is welcome.

**Comment by dvandersluis** [comment]
My apologies, I missed that line. I don't have a strong feeling either way, just by looking at the benchmark that was my immediate thought.

File: lib/rubocop/cop/style/concat_array_literals.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      # to avoid redundant array literals.
+      #
+      # @safety
+      #   This cop is unsafe, as it can produce false positives if the receiver
+      #   is not an `Array` object.
+      #
+      # @example
+      #
+      #   # bad
+      #   list.concat([item])
```
Comment (bbatsov) [inline lib/rubocop/cop/style/concat_array_literals.rb:16]: I think it's better to add examples with multiple elements as well, so people won't think this works only one-element concats.

Comment (koic) [reply lib/rubocop/cop/style/concat_array_literals.rb]: Exactly! I've added examples.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**